File: test/language/builtin_types/cpp/BitFieldUInt64LengthTest.cpp
#include <limits>

#include "gtest/gtest.h"

#include "builtin_types/bitfield_uint64_length/Container.h"

#include "zserio/SerializeUtil.h"

namespace builtin_types
{
namespace bitfield_uint64_length
{

TEST(BitFieldUIn64LengthTest, bitSizeOf)
{
    Container container;
    const uint64_t bitFieldLength = 33;
    container.setLength(bitFieldLength);
    container.setUnsignedBitField(std::numeric_limits<uint32_t>::max() + static_cast<uint64_t>(1));
    container.setSignedBitField(std::numeric_limits<int32_t>::max() + static_cast<int64_t>(1));
    const size_t expectedBitSizeOfContainer = 64 + 33 + 33;
    ASSERT_EQ(expectedBitSizeOfContainer, container.bitSizeOf());
}

TEST(BitFieldUIn64LengthTest, readWrite)
{
    Container container;
    const uint64_t bitFieldLength = 33;
    container.setLength(bitFieldLength);
    container.setUnsignedBitField(std::numeric_limits<uint32_t>::max() + static_cast<uint64_t>(1));
    container.setSignedBitField(std::numeric_limits<int32_t>::max() + static_cast<int64_t>(1));

    const std::string fileName = "language/builtin_types/bit_field_uint64_length.blob";
    zserio::serializeToFile(container, fileName);

    const Container readContainer = zserio::deserializeFromFile<Container>(fileName);
    ASSERT_TRUE(container == readContainer);
}

} // namespace bitfield_uint64_length
} // namespace builtin_types

File: test/language/builtin_types/python/BitFieldUInt64LengthTest.py
import os
import zserio

import BuiltInTypes

from testutils import getApiDir


class BitFieldUInt64LengthTest(BuiltInTypes.TestCase):
    def testBitSizeOf(self):
        container = self.api.Container()
        bitFieldLength = 33
        container.length = bitFieldLength
        container.unsigned_bit_field = zserio.limits.UINT32_MAX + 1
        container.signed_bit_field = zserio.limits.INT32_MAX + 1

        expectedBitSizeOfContainer = 64 + 33 + 33
        self.assertEqual(expectedBitSizeOfContainer, container.bitsizeof())

    def testReadWrite(self):
        container = self.api.Container()
        bitFieldLength = 33
        container.length = bitFieldLength
        container.unsigned_bit_field = zserio.limits.UINT32_MAX + 1
        container.signed_bit_field = zserio.limits.INT32_MAX + 1

        zserio.serialize_to_file(container, self.BLOB_NAME)

        readContainer = zserio.deserialize_from_file(self.api.Container, self.BLOB_NAME)
        self.assertEqual(container, readContainer)

    BLOB_NAME = os.path.join(getApiDir(os.path.dirname(__file__)), "bit_field_uint64_length.blob")
